### PR TITLE
Move OIDCSessionParticipantCache to tenant space

### PR DESCRIPTION
**Pull request by UdeshAthukorala**
### Proposed changes in this pull request

Move the following caches to the tenant space.

- OIDCSessionParticipantCache

Related Issue : wso2/product-is#11690

### CODE REVIEW

File: components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCLogoutServlet.java [hunk 1]
```diff
@@ -973,4 +975,16 @@ private void clearTokenBindingElements(String clientId, HttpServletRequest reque
         tokenBinders.stream().filter(t -> oAuthAppDO.getTokenBindingType().equals(t.getBindingType())).findAny()
                 .ifPresent(t -> t.clearTokenBindingElements(request, response));
     }
+
+    private String resolveTenantDomain(HttpServletRequest request) {
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCLogoutServlet.java:979]: So the idea hear is if the tenanted sessions are not enabled, cache will be in the super tenant?

File: components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/cache/OIDCSessionParticipantCache.java [hunk 2]
```diff
@@ -66,9 +66,10 @@ public static OIDCSessionParticipantCache getInstance() {
      * @param entry Actual object where cache entry is placed.
      */
     @Override
-    public void addToCache(OIDCSessionParticipantCacheKey key, OIDCSessionParticipantCacheEntry entry) {
+    public void addToCache(OIDCSessionParticipantCacheKey key, OIDCSessionParticipantCacheEntry entry,
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/cache/OIDCSessionParticipantCache.java:69]: Is there any thing we can do about this API change?

File: components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCLogoutServlet.java (hunk 1 above)
Comment (UdeshAthukorala) [reply components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCLogoutServlet.java:979]: Yes. If the tenanted sessions are not enabled, cache will be added to the super tenant.

**Review by senthalan — APPROVED** [review]

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dto/OAuth2AuthorizeReqDTO.java [hunk 3]
```diff
@@ -232,4 +234,12 @@ public String getIdpSessionIdentifier() {
     public void setIdpSessionIdentifier(String idpSessionIdentifier) {
         this.idpSessionIdentifier = idpSessionIdentifier;
     }
+
+    public String getLoginTenantDomain() {
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dto/OAuth2AuthorizeReqDTO.java:238]: getLoggedInTenenatDomain makes much sense right?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dto/OAuth2AuthorizeReqDTO.java [hunk 4]
```diff
@@ -53,6 +53,8 @@
     private String sessionDataKey;
     // Identifier of the authenticated framework session. This will be included as a claim in the ID token.
     private String idpSessionIdentifier;
+    // Set the login tenant domain.
+    private String loginTenantDomain;
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dto/OAuth2AuthorizeReqDTO.java:57]: How about LoggedInTenenatDomain?

File: components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/OIDCSessionManager.java [hunk 5]
```diff
@@ -32,69 +34,167 @@
...
+    public boolean sessionExists(String sessionId, String loginTenantDomain) {
+
+        String tenantDomain = resolveCacheTenantDomain(loginTenantDomain);
+        return getOIDCSessionState(sessionId, tenantDomain) != null;
+    }
+
+    private String resolveCacheTenantDomain(String tenantDomain) {
 
-        return getOIDCSessionState(sessionId) != null;
+        if (!IdentityTenantUtil.isTenantedSessionsEnabled()) {
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/OIDCSessionManager.java:195]: Its good to have explained if-else clause but here can't we use one-line logic using ? and :

File: components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/backchannellogout/LogoutRequestSender.java [hunk 6]
```diff
@@ -88,10 +93,26 @@ public void sendLogoutRequests(HttpServletRequest request) {
...
+    @Deprecated
     public void sendLogoutRequests(String opbsCookieId) {
 
-        Map<String, String> logoutTokenList = getLogoutTokenList(opbsCookieId);
+        // For backward compatibility, SUPER_TENANT_DOMAIN was added as the cache maintained tenant.
+        sendLogoutRequests(opbsCookieId, MultitenantConstants.SUPER_TENANT_DOMAIN_NAME);
+    }
+
+    /**
+     * Sends logout requests to all service providers.
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/backchannellogout/LogoutRequestSender.java:108]: All sp means all sp in particular tenant-right?.... How we handle SAAS applications?

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 7]
```diff
@@ -2487,6 +2486,7 @@ private OAuth2AuthorizeReqDTO buildAuthRequest(OAuth2Parameters oauth2Params, Se
         authzReqDTO.setEssentialClaims(oauth2Params.getEssentialClaims());
         authzReqDTO.setSessionDataKey(oauth2Params.getSessionDataKey());
         authzReqDTO.setIdpSessionIdentifier(sessionDataCacheEntry.getSessionContextIdentifier());
+        authzReqDTO.setLoginTenantDomain(oauth2Params.getLoginTenantDomain());
```
Comment (jeradrutnam) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2489]: Need a space before the `If`

File: components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCLogoutServlet.java [hunk 8]
```diff
@@ -973,4 +975,16 @@ private void clearTokenBindingElements(String clientId, HttpServletRequest reque
         tokenBinders.stream().filter(t -> oAuthAppDO.getTokenBindingType().equals(t.getBindingType())).findAny()
                 .ifPresent(t -> t.clearTokenBindingElements(request, response));
     }
+
```
[the remark this answers is not in the document]
Comment (UdeshAthukorala) [reply components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCLogoutServlet.java:978]: As far as I know their is no such utill method to resolve tenant domain.

File: components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCLogoutServlet.java [hunk 9]
```diff
@@ -857,9 +859,9 @@ private boolean getOpenIDConnectSkipUserConsent(HttpServletRequest request) thro
      *
      * @param opbsCookieValue OP browser state cookie value.
      */
-    private void doBackChannelLogout(String opbsCookieValue) {
+    private void doBackChannelLogout(String opbsCookieValue, String tenantDomain) {
```
[the remark this answers is not in the document]
Comment (UdeshAthukorala) [reply components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCLogoutServlet.java:862]: Addressed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1705

File: components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/handler/OIDCLogoutEventHandler.java [hunk 10]
```diff
@@ -164,4 +167,19 @@ private boolean hasOPBSCookieValue(Cookie opbsCookie) {
 
         return StringUtils.isNotBlank(opbsCookieValue);
     }
+
+    private String resolveTenantDomain(HttpServletRequest request) {
```
[the remark this answers is not in the document]
Comment (UdeshAthukorala) [reply components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/handler/OIDCLogoutEventHandler.java:171]: Addressed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1705

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dto/OAuth2AuthorizeReqDTO.java (hunk 3 above)
Comment (UdeshAthukorala) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dto/OAuth2AuthorizeReqDTO.java:238]: Addressed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1705

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dto/OAuth2AuthorizeReqDTO.java (hunk 4 above)
Comment (UdeshAthukorala) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dto/OAuth2AuthorizeReqDTO.java:57]: Addressed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1705

File: components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/OIDCSessionManager.java (hunk 5 above)
Comment (UdeshAthukorala) [reply components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/OIDCSessionManager.java:195]: For the readability purposes i think it is better to have this explained if-else clause

File: components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/backchannellogout/LogoutRequestSender.java (hunk 6 above)
Comment (UdeshAthukorala) [reply components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/backchannellogout/LogoutRequestSender.java:108]: Yes. For both sass and non-saas applications we used this method.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 7 above)
Comment (UdeshAthukorala) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2489]: Addressed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1705

**Pull request merged**